Extend HTTP client tracing by message body

DDvO opened the pull request:
* add `OSSL_TRACE_STRING`, `OSSL_TRACE_STRING_MAX`, and `OSSL_trace_string()`
* `OSSL_HTTP_REQ_CTX_nbio()`: use `OSSL_TRACE_STRING()` for msg body where it makes sense

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

DDvO commented:
Ping for reviews

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

wan7teen-ping commented:
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Sab, 8 Okt 2022, 8:09 AM openssl-machine ***@***.***>
menulis:

> This PR is in a state where it requires action by @openssl/otc
> <https://github.com/orgs/openssl/teams/otc> but the last update was 30
> days ago
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openssl/openssl/pull/18704#issuecomment-1272171037>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APESUCRGXFJ75YHLA2GJQSLWCC3T3ANCNFSM52NX6WDQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

paulidale commented on crypto/trace.c:
```diff
@@ -519,3 +519,26 @@ void OSSL_trace_end(int category, BIO * channel)
+    }
+    if (!text) { /* mask control characters */
+        for (i = 0; i < len; i++, data++)
+            buf[i] = (char)*data < ' ' ? ' ' : *data;
```
127 is normally considered a control character too.
iscntrl() or ossl_iscntrl() perhaps?

DDvO replied:
Good point - now using `ossl_iscntrl()`
and excempting the `\n` case in order to preserve existing newlines.

DDvO commented:
Also rebased to fix trivial merge conflicts in `libcrypto.num` due to 3.1->3.2 changes

DDvO commented:
Rebased to fix a new trivial merge conflict.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO commented:
Ping for 2nd review

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

t8m commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -1067,9 +1067,9 @@ although they usually contain hints that would be helpful for diagnostics.
+If OpenSSL was built with trace support enabled (e.g., C<./config enable-trace>)
 and the environment variable B<OPENSSL_TRACE> includes B<HTTP>,
-the request and response headers of HTTP transfers are printed.
+the requests and response headers of transferred via HTTP are printed.
```
The `of` seems superfluous?

DDvO replied:
Right, so removed it.

t8m commented on doc/man3/OSSL_trace_enabled.pod:
```diff
@@ -287,6 +309,9 @@ L<OSSL_trace_set_channel(3)>, L<OSSL_trace_set_callback(3)>
 The OpenSSL Tracing API was added in OpenSSL 3.0.
 
+OSSL_TRACE_STRING(), OSSL_TRACE_STRING_MAX, and OSSL_trace_string
+were added in OpensSL 3.1.
```
`OpenSSL 3.2`

DDvO replied:
Updated.

t8m reviewed (changes requested):
Would it be possible to add some tests or is too big effort to add them?

DDvO commented:
Good point to add tests - done.
Doing so, I found a mem management bug and doc omission, fixed in a separate PR.
BTW, renamed `90-test_traceapi.t` to `90-test_trace_api.t` for consistency.

I had to rebase in order to fix merge conflicts.

t8m commented on doc/man3/OSSL_trace_enabled.pod:
```diff
@@ -287,6 +310,9 @@ L<OSSL_trace_set_channel(3)>, L<OSSL_trace_set_callback(3)>
 The OpenSSL Tracing API was added in OpenSSL 3.0.
 
+OSSL_TRACE_STRING(), OSSL_TRACE_STRING_MAX, and OSSL_trace_string
+were added in OpensSL 3.2.
```
Nit: OpenSSL

t8m commented on test/trace_api_test.c:
```diff
@@ -65,18 +65,45 @@ static int test_trace_categories(void)
+#define OSSL_END "-abc"
+
+#define trace_string(text, full, str) \
+    OSSL_trace_string(trc_out, text, full, (unsigned char*)(str), strlen(str))
```
Nit: space between char and `*`

DDvO commented:
Thanks for the immediate further comments.
Fixed both nits, as well as a mistake on indentation nesting of `#define`.

DDvO commented:
> CI is relevant.

Doc nit fixed.

t8m commented:
@paulidale please reconfirm

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Merged - thanks @t8m and @paulidale for your comments and approvals.
Glad that this is through now.
